src/fairseq2/recipes/hg/asr_eval.py

```diff
@@ -17,44 +18,86 @@
 )
 
 from fairseq2.config_registry import ConfigRegistry
-from fairseq2.data.data_pipeline import SequenceData
 from fairseq2.data.text import load_text_tokenizer
 from fairseq2.data.text.text_tokenizer import TextTokenizer
 from fairseq2.datasets.batching import StaticBatching
 from fairseq2.logging import get_log_writer
-from fairseq2.models.seq2seq import Seq2SeqBatch
 from fairseq2.models.sequence import SequenceBatch
 from fairseq2.models.wav2vec2.asr import load_wav2vec2_asr_model
-from fairseq2.nn.padding import get_seqs_and_padding_mask
+from fairseq2.nn.padding import PaddingMask, get_seqs_and_padding_mask
 from fairseq2.recipes.hg.dataset import Example, create_hf_reader
 from fairseq2.recipes.hg.evaluator import HFEvaluator
 from fairseq2.recipes.utils.setup import setup_root_gang
-from fairseq2.typing import META, DataType
+from fairseq2.typing import META, DataType, Device
 from fairseq2.utils.profiler import Stopwatch
 
 log = get_log_writer(__name__)
 
 
+@dataclass
+class AsrDatasetConfig:
+    """Configuration for an automatic speech recognition dataset."""
+
+    dataset_path: str
+    """The name of the dataset."""
+
+    dataset_name: Optional[str] = None
+    """The name of the dataset split."""
+
+    split: str = "test"
+    """The name of the dataset split to evaluate with."""
+
+    source_column: List[str] = field(default_factory=list)
```

Comment:
If we expect this to accept multiple column names, I would prefer `source_columns` and `target_columns`.

Reply:
Not exactly. The `source_column` and `target_column` parameters define the paths to the source and target data within the dataset structure.

Consider the following dataset structures:

- **LibriSpeech:**
    ```json
    dataset: {
        audio: {
            array: [AUDIO_WAVE_DATA]
        },
        text: [TEXT_DATA]
    }
    ```
    - Source data: `dataset['audio']['array']`
    - Target data: `dataset['text']`

- **Google/Fleurs:**
    ```json
    dataset: {
        audio: {
            array: [AUDIO_WAVE_DATA]
        },
        transcriptions: [TEXT_DATA]
    }
    ```
    - Source data: `dataset['audio']['array']`
    - Target data: `dataset['transcriptions']`

- **Another Dataset:**
    ```json
    dataset: {
        audio: {
            waveform: {
                array: [AUDIO_WAVE_DATA]
            },
            other_format: {
                array: [AUDIO_WAVE_DATA]
            }
        },
        text: {
            raw: {
                data: [TEXT_DATA]
            }
        }
    }
    ```
    - Source data: `dataset['audio']['waveform']['array']`
    - Target data: `dataset['text']['raw']['data']`

To make the code adaptable to different datasets, regardless of their structure, the `AsrDatasetConfig` can be used to specify the paths to the source and target data as lists. This allows for greater flexibility and compatibility across various dataset formats.

For example:

- **Dataset 1:**
    - Source: `['audio', 'array']`
    - Target: `['text']`
  
- **Dataset 2:**
    - Source: `['audio', 'array']`
    - Target: `['transcriptions']`
  
- **Dataset 3:**
    - Source: `['audio', 'waveform', 'array']`
    - Target: `['text', 'raw', 'data']`

Is there a better approach to handle this? Should I rename the fields to `source_path` and `target_path` instead?

Reply:
In https://github.com/facebookresearch/fairseq2/pull/749/commits/667194fd968c82f46ff43e85aaf57f13127e3be7, I have renamed the fields to `source_column_path` and `target_column_path`

I'll mark the comment as resolved, but please let me know if this change doesn't align with your expectations.